### DISCUSSION TITLE
Odd behavior of physics joints when hundreds of static bodies are created at runtime, and scene is reloaded.

### DISCUSSION THREAD

**Discussion by addmix** [godot-jolt/godot-jolt#864] [Questions] [answered]
This an odd and very specific issue I have been having for quite a while. I am opting to open a discussion here, rather than opening a ticket because I have not been able to create a minimal reproduction. I have however, found that it is being caused (in my case) by an interaction between StaticBody3D and Generic6DOFJoints when a scene is loaded or reloaded. If I had to make a guess on why it is happening, is that the unique IDs or related state for the static bodies and joints are not freed after the nodes are reloaded. I assume that this would be replicable using any PhysicsBodies and PhysicsJoints.

 I have missiles attached to an airplane using Generic6DOFJoints. After the scene has been reloaded twice (using `get_tree().reload_current_scene()`), the joints seem to solve for the wrong positions, or otherwise not respect the physics simulation directly.

Here's a video of the problem:
https://youtu.be/OvAAw6WcZ9s

Here's a video of the same demo but using Godot Physics, no issue present:
https://youtu.be/gFhIN-9dnRM

The way I produce this issue, I run my game, reload the scene twice, and when the airplane gains velocity, the missiles shift to a position ahead of themselves, proportional to their linear_velocity. Every time the game window is closed, and re-run, this issue is always exhibited on the 2nd reload of the scene. The specific missiles that exhibit the behavior is random, but the amount of missiles that exhibit behavior seems to be correlated to the amount of staticbodies used by the terrain system.

I have found that this behavior is in some way caused by the terrain's staticbodies. If I disable the static bodies from being added to the scene, the behavior returns to normal. If I set the static body's process_mode to Node.PROCESS_MODE_DISABLED, behavior is normal. If I disable the static bodies collision shapes, the behavior remains.

The terrain being used is something I have created, it is a separate plugin and in no way interfaces anywhere else in the project. It uses threads and compute shaders for terrain gen. There are about 400 static bodies generated by the terrain. These nodes are not serialized, and are generated during the first few seconds after the scene has been opened. This is the code used on the terrain chunks to add the static bodies:
```
func _ready() -> void:
	add_children)

func add_children() -> void:
	add_child(mesh_instance, false, Node.INTERNAL_MODE_FRONT)
	if not terrain.skip_collision_generation:
		add_child(static_body, false, Node.INTERNAL_MODE_FRONT)
```

Here is the node layout of the missiles:
All that the script on the Generic6DOFJoint does, is finds a child node (the missile), and sets it as the Joint's Node B. I did test a different node layout, with the missiles being siblings to the Joint, but the behavior was the same.
![image](https://github.com/godot-jolt/godot-jolt/assets/38119387/f3357766-814c-4467-9989-31d99b52dd00)

I have tried increasing Jolt's max bodies setting, but it had no effect. In my attempts to make a minimal reproduction project, I tried instancing hundreds/thousands of static bodies at runtime, but I was not able to recreate the issue.

**Comment by mihe** [comment]
Are those missiles (or their joints) set as `top_level` or are they actually childed to the airplane?

**Reply by addmix** [reply]
I included a screenshot in the original message. They are direct children, Airplane > 6DOF > Missile

**Reply by mihe** [reply, marked as answer]
A `RigidBody3D` (which I'm assuming your missiles are?) should ideally never be a descendant of another moving node, at least not one that moves every frame, like your airplane. This will cause the missiles to have their transform set every frame, which will interfere with the physics simulation. This is a problem in Godot Physics as well as Godot Jolt, although the ill effects with Jolt are typically worse and more pronounced, and is possibly what's causing these weird issues with the missiles.

I realize this may sound like a frustrating limitation, but that's where the [`Node3D.top_level`](https://docs.godotengine.org/en/stable/classes/class_node3d.html#class-node3d-property-top-level) property comes in handy, as you can have your desired scene structure and then right before the simulation starts you set this to `true` in order to detach the body from its hierarchy.

So I would suggest trying this: Override the `_init()` method of your missiles and simply do `top_level = true` in there, and see if that changes anything.

You likely don't want to just set it in the editor, as `top_level` also (unfortunately) affects the body when moving its ancestors around with the move tool in the editor.

**Reply by mihe** [reply]
It should also be noted that if you don't necessarily need the behavior of the joints for this, you could just as well skip the joint and instead set `freeze_mode` to `FREEZE_MODE_KINEMATIC` and then `freeze` to `true` for the duration of the time that the missile is attached to the airplane, both of which you can set through the editor just fine. Then once you release the missile you similarly set its `top_level` to `true` and its `freeze` to `false` through the script.

Kinematic bodies have no problems being childed to another moving node.

That will likely be a bit cheaper computationally speaking, and likely less prone to jank as well.

**Reply by addmix** [reply]
Using `top_level` adds a little bit of complexity, as my game uses a floating origin, and adds one more thing for me to (forget to) configure, but I've managed to get it working as expected. Though, I still do think it's peculiar that the "ghost connection" between the airplane and the missiles only happened if the scene was reloaded multiple times, and there were some amount of staticbodies in the scene, something I have yet to replicate when using GodotPhysics.

**Reply by mihe** [reply]
Glad to hear it worked.

Yeah, I can't explain the randomness of it. The running theory last time something like this was brought up was that some tiny error is randomly (but deterministically) introduced into the feedback loop of transforms that then gets amplified as time goes on. Either way, the symptoms through which this manifests is likely to be specific to each engine.

**Comment by addmix** [comment]
After doing more testing, and removing more and more unrelated features and code (not finished yet), I did manage to confirm one of my hunches. It seems that the missiles forget that they are descendants of the airplane. Haven't managed to track down what exactly has caused this in the first place. I do have another hunch that the issue could be caused by the "inherited scene" feature of Godot.
https://youtu.be/MwrxmOuoPfs
